### PR TITLE
Add support for dropping a database that has any table that has any reference column

### LINKED ISSUE #390: 外部キー制約が存在するときDROP DATABASEが失敗する
外部キー制約が存在するとき、`ENGINE=InnoDB` では実行可能なDROP DATABASEが、`ENGINE=Mroonga` では失敗するようです。

以下はDockerイメージ「groonga/mroonga:mysql5733_mroonga1011」で実行した結果です。

# `ENGINE=InnoDB` の場合

SQL:

```sql
-- インデックスが存在する場合にDROP DATABASEが失敗しないように
DROP TABLE IF EXISTS innodb_drop_database_test.t2;

DROP DATABASE IF EXISTS innodb_drop_database_test;

CREATE DATABASE innodb_drop_database_test;
USE innodb_drop_database_test;

CREATE TABLE t1 (
  id int PRIMARY KEY AUTO_INCREMENT
) ENGINE=InnoDB DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;

CREATE TABLE t2 (
  id int PRIMARY KEY AUTO_INCREMENT,
  id_t1 int NOT NULL,
  CONSTRAINT `id_t1` FOREIGN KEY (`id_t1`) REFERENCES `t1` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;

SHOW CREATE TABLE t1;
SHOW CREATE TABLE t2;

DROP DATABASE innodb_drop_database_test;
```

実行結果（正常終了）:

```
Table	Create Table
t1	CREATE TABLE `t1` (\n  `id` int(11) NOT NULL AUTO_INCREMENT,\n  PRIMARY KEY (`id`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 ROW_FORMAT=DYNAMIC
Table	Create Table
t2	CREATE TABLE `t2` (\n  `id` int(11) NOT NULL AUTO_INCREMENT,\n  `id_t1` int(11) NOT NULL,\n  PRIMARY KEY (`id`),\n  KEY `id_t1` (`id_t1`),\n  CONSTRAINT `id_t1` FOREIGN KEY (`id_t1`) REFERENCES `t1` (`id`)\n) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 ROW_FORMAT=DYNAMIC
```

# `ENGINE=Mroonga` の場合

SQL:

```sql
-- インデックスが存在する場合にDROP DATABASEが失敗しないように
DROP TABLE IF EXISTS mroonga_drop_database_test.t2;

DROP DATABASE IF EXISTS mroonga_drop_database_test;

CREATE DATABASE mroonga_drop_database_test;
USE mroonga_drop_database_test;

CREATE TABLE t1 (
  id int PRIMARY KEY AUTO_INCREMENT
) ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;

CREATE TABLE t2 (
  id int PRIMARY KEY AUTO_INCREMENT,
  id_t1 int NOT NULL,
  CONSTRAINT `id_t1` FOREIGN KEY (`id_t1`) REFERENCES `t1` (`id`)
) ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;

SHOW CREATE TABLE t1;
SHOW CREATE TABLE t2;

DROP DATABASE mroonga_drop_database_test;
```

実行結果（エラー）:

以下のメッセージが表示され、テーブル `t1` のみ残ります。

```
Table	Create Table
t1	CREATE TABLE `t1` (\n  `id` int(11) NOT NULL AUTO_INCREMENT,\n  PRIMARY KEY (`id`)\n) ENGINE=Mroonga DEFAULT CHARSET=utf8mb4 ROW_FORMAT=DYNAMIC
Table	Create Table
t2	CREATE TABLE `t2` (\n  `id` int(11) NOT NULL AUTO_INCREMENT,\n  `id_t1` int(11) NOT NULL,\n  PRIMARY KEY (`id`),\n  KEY `id_t1` (`id_t1`),\n  CONSTRAINT `id_t1` FOREIGN KEY (`id_t1`) REFERENCES `mroonga_drop_database_test`.`t1` (`id`) ON DELETE RESTRICT ON UPDATE RESTRICT\n) ENGINE=Mroonga DEFAULT CHARSET=utf8mb4 ROW_FORMAT=DYNAMIC
ERROR 1016 (HY000) at line 22: [table][remove] a column that references the table exists: <t2.id_t1> -> <t1>
```

### PR DESCRIPTION
**Pull request by komainu8**
GitHub: fix #390
Reported by ochaochaocha3. Thanks!!!

### CODE REVIEW

**Comment by kou** [comment]
Does this accept `DROP TABLE t1` before `DROP DATABASE another`? It breaks `t2` because `t2.id_t1` is removed.
(InnoDB doesn't accept `DROP TABLE t1`.)

**Comment by komainu8** [comment]
Ah...
This is not accept `DROP TABLE t1` before `DROP DATABASE` another.

I'll fix it.

File: ha_mroonga.cpp [hunk 1]
```diff
@@ -5761,7 +5761,12 @@ int ha_mroonga::generic_delete_table(const char *name,
   error = drop_indexes(table_name);
   grn_obj *table_obj = grn_ctx_get(ctx, table_name, strlen(table_name));
   if (table_obj) {
-    grn_obj_remove(ctx, table_obj);
+    if ((thd_sql_command(ha_thd()) == SQLCOM_DROP_DB) &&
+        have_reference_column(table_obj)) {
```
Comment (kou) [inline ha_mroonga.cpp:5765]: Do we still need this check?

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+--disable_warnings
+DROP DATABASE IF EXISTS another;
+--enable_warnings
+
+CREATE DATABASE another;
+USE another;
+
+CREATE TABLE t1 (
+  id int PRIMARY KEY AUTO_INCREMENT
+) ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test:30]: Is `ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC` needed?

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+--source ../../../../include/mroonga/have_mroonga.inc
+
+SET NAMES UTF8MB4;
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test:19]: Is this needed?

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+--disable_warnings
+DROP DATABASE IF EXISTS another;
+--enable_warnings
+
+CREATE DATABASE another;
+USE another;
+
+CREATE TABLE t1 (
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test:28]: Could you use more descriptive name like `referred` for `t1` and `refer` for `t2`?

File: mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  id_t1 int NOT NULL,
+  CONSTRAINT `id_t1` FOREIGN KEY (`id_t1`) REFERENCES `t1` (`id`)
+) ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;
+
+--error ER_CANT_OPEN_FILE
+DROP TABLE t1;
+
+ALTER TABLE t2 DROP FOREIGN KEY `id_t1`;
+DROP TABLE t1;
+DROP TABLE t2;
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test:43]:
```suggestion
DROP TABLE t2;
DROP TABLE t1;
```

File: mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+--source ../../../../include/mroonga/have_mroonga.inc
+
+SET NAMES UTF8MB4;
+
+--disable_warnings
+DROP DATABASE IF EXISTS test;
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test:22]: Why should we need to recreate `test` database?

File: ha_mroonga.cpp (hunk 1 above)
Comment (komainu8) [reply ha_mroonga.cpp]: This is needless.
I've removed it.

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test (hunk 2 above)
Comment (komainu8) [reply mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test]: This is needless.
I've removed it.

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test (hunk 3 above)
Comment (komainu8) [reply mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test]: This is needless.
I've removed it.

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test (hunk 4 above)
Comment (komainu8) [reply mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test]: I've renamed t1 to referred and t2 to refer.

File: mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test (hunk 6 above)
Comment (komainu8) [reply mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test]: I've made a mistake.
I removed recreation `test` database.

File: ha_mroonga.hpp [hunk 7]
```diff
@@ -861,6 +861,7 @@ class ha_mroonga: public handler
   ulonglong file_size(const char *path);
 
   bool have_unique_index();
+  bool have_reference_column(grn_obj *table);
```
Comment (kou) [inline ha_mroonga.hpp:864]: Could you remove needless code?

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+CREATE TABLE referred (
+  id int PRIMARY KEY AUTO_INCREMENT
+);
+
+CREATE TABLE refer (
+  id int PRIMARY KEY AUTO_INCREMENT,
+  id_referred int NOT NULL,
+  CONSTRAINT `id_referred` FOREIGN KEY (`id_referred`) REFERENCES `referred` (`id`)
+) ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test:34]: ` ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC` is needless.

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+USE another;
+
+CREATE TABLE referred (
+  id int PRIMARY KEY AUTO_INCREMENT
+);
+
+CREATE TABLE refer (
+  id int PRIMARY KEY AUTO_INCREMENT,
+  id_referred int NOT NULL,
+  CONSTRAINT `id_referred` FOREIGN KEY (`id_referred`) REFERENCES `referred` (`id`)
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test:33]:
```suggestion
  CONSTRAINT id_referred FOREIGN KEY (id_referred) REFERENCES referred (id)
```

File: mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+
+--source ../../../../include/mroonga/have_mroonga.inc
+
+USE test;
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test:19]: Is this needed?

File: mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+) ENGINE=mroonga DEFAULT CHARSET utf8mb4 ROW_FORMAT=DYNAMIC;
+
+DROP DATABASE another;
+
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/t/have_foreign_key.test:37]: `USE test` is missing. `test` database is used by default in test.

File: mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test:20]: `DROP TABLE IF EXISTS`s are missing.

File: mysql-test/mroonga/storage/drop/database/r/have_foreign_key.result [hunk 13]
```diff
@@ -0,0 +1,13 @@
+DROP DATABASE IF EXISTS another;
+CREATE DATABASE another;
+USE another;
+CREATE TABLE referred (
+id int PRIMARY KEY AUTO_INCREMENT
+);
+CREATE TABLE refer (
+id int PRIMARY KEY AUTO_INCREMENT,
+id_referred int NOT NULL,
+CONSTRAINT `id_referred` FOREIGN KEY (`id_referred`) REFERENCES `referred` (`id`)
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/database/r/have_foreign_key.result:10]:
```suggestion
CONSTRAINT id_referred FOREIGN KEY (id_referred) REFERENCES referred (id)
```

File: mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+--enable_warnings
+
+CREATE TABLE referred (
+  id int PRIMARY KEY AUTO_INCREMENT
+);
+
+CREATE TABLE refer (
+  id int PRIMARY KEY AUTO_INCREMENT,
+  id_referred int NOT NULL,
+  CONSTRAINT `id_referred` FOREIGN KEY (`id_referred`) REFERENCES `referred` (`id`)
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/table/t/have_foreign_key.test:30]:
```suggestion
  CONSTRAINT id_referred FOREIGN KEY (id_referred) REFERENCES referred (id)
```

File: mysql-test/mroonga/storage/drop/table/r/have_foreign_key.result [hunk 15]
```diff
@@ -0,0 +1,13 @@
+DROP TABLE IF EXISTS referred, refer;
+CREATE TABLE referred (
+id int PRIMARY KEY AUTO_INCREMENT
+);
+CREATE TABLE refer (
+id int PRIMARY KEY AUTO_INCREMENT,
+id_referred int NOT NULL,
+CONSTRAINT `id_referred` FOREIGN KEY (`id_referred`) REFERENCES `referred` (`id`)
```
Comment (kou) [inline mysql-test/mroonga/storage/drop/table/r/have_foreign_key.result:8]:
```suggestion
CONSTRAINT id_referred FOREIGN KEY (id_referred) REFERENCES referred (id)
```

**Pull request merged**